File: cpp/linkedList/lru_cache.cpp
#include <bits/stdc++.h>
using namespace std;
#define ll long long
#define pb push_back
#define all(x) x.begin(), x.end()
#define endl '\n'

class LRUCache {
  public:
    struct Node {
        int key, val;
        Node* prev;
        Node* next;
        Node(int k, int v)
          : key(k)
          , val(v)
          , prev(nullptr)
          , next(nullptr) {
        }
    };

    unordered_map<int, Node*> cache;
    int cap;
    int currentCap;
    Node* head;
    Node* tail;

    LRUCache(int capacity)
      : cap(capacity)
      , currentCap(0) {
        head = new Node(0, 0);
        tail = new Node(0, 0);
        head->next = tail;
        tail->prev = head;
    }

    int get(int key) {
        if (cache.find(key) == cache.end()) return -1;
        Node* curr = cache[key];
        curr->prev->next = curr->next;
        curr->next->prev = curr->prev;
        curr->next = tail;
        curr->prev = tail->prev;
        tail->prev->next = curr;
        tail->prev = curr;
        return curr->val;
    }

    void put(int key, int value) {
        if (cache.find(key) != cache.end()) {
            Node* curr = cache[key];
            curr->val = value;
            curr->prev->next = curr->next;
            curr->next->prev = curr->prev;
            curr->next = tail;
            curr->prev = tail->prev;
            tail->prev->next = curr;
            tail->prev = curr;
        } else {
            if (currentCap == cap) {
                Node* lru = head->next;
                cache.erase(lru->key);
                head->next = lru->next;
                lru->next->prev = head;
                delete lru;
                currentCap--;
            }
            Node* curr = new Node(key, value);
            curr->next = head->next;
            curr->prev = head;
            head->next->prev = curr;
            head->next = curr;
            cache[key] = curr;
            currentCap++;
        }
    }
};

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    return 0;
}

File: python/linkdedList/lruCache.py
import threading


class LRUCache:

    class ListNode:
        def __init__(self, val=0, next=None, prev=None, key=0):
            self.val = val
            self.next = next
            self.prev = prev
            self.key = key

    def __init__(self, capacity: int):
        self.node_map = {}
        self.head = self.ListNode()
        self.tail = self.ListNode()
        self.current_capacity = 0
        self.capacity = capacity
        self.head.next = self.tail
        self.tail.prev = self.head

    def get(self, key: int) -> int:
        if key not in self.node_map:
            return -1

        current = self.node_map[key]
        current.prev.next = current.next
        current.next.prev = current.prev
        current.next = self.head.next
        current.prev = self.head
        self.head.next.prev = current
        self.head.next = current
        return current.val

    def put(self, key: int, value: int) -> None:
        if key in self.node_map:
            self.node_map[key].val = value
            self.get(key)
            return

        current = self.ListNode(value, key=key)
        current.next = self.head.next
        current.prev = self.head
        self.head.next.prev = current
        self.head.next = current
        self.node_map[key] = current

        if self.current_capacity < self.capacity:
            self.current_capacity += 1
        else:
            lru = self.tail.prev
            lru.prev.next = self.tail
            self.tail.prev = lru.prev
            del self.node_map[lru.key]


def main():
    cache = LRUCache(2)
    cache.put(1, 1)
    cache.put(2, 2)
    cache.put(3, 3)
    print(cache.node_map.keys())


if __name__ == "__main__":
    threading.Thread(target=main).start()
